src/main/java/org/apache/sling/resourceresolver/impl/FactoryRegistrationHandler.java

```diff
@@ -0,0 +1,178 @@
+                    final Runnable unregisterTask = runnables.get(runnables.size() - 2);
+                    unregisterTask.run();
+                }
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    /**
+     * Check the preconditions and if it changed, either register factory or unregister
+     */
+    void maybeRegisterFactory(final String unavailableName, final String unavailableServicePid) {
+        if (!factoryRegistrationWorker.isShutdown()) {
+            LOG.debug("submitting maybeRegisterFactory");
+            factoryRegistrationWorker.submit(() -> {
+                final boolean preconditionsOk = factoryPreconditions.checkPreconditions(unavailableName, unavailableServicePid);
+                if ( preconditionsOk && this.factoryRegistration == null ) {
+                    // check system bundle state - if stopping, don't register new factory
+                    final Bundle systemBundle = activator.getBundleContext().getBundle(Constants.SYSTEM_BUNDLE_LOCATION);
+                    if ( systemBundle != null && systemBundle.getState() != Bundle.STOPPING ) {
+                        withThreadName("ResourceResolverFactory registration", this::doRegisterFactory);
+                    }
+                } else if ( !preconditionsOk && this.factoryRegistration != null ) {
+                    LOG.debug("performing unregisterFactory via maybeRegisterFactory");
+                    withThreadName("ResourceResolverFactory deregistration", this::doUnregisterFactory);
+                }
+            });
+        }
+    }
+
+    void unregisterFactory() {
+        LOG.debug("submitting unregisterFactory");
+        factoryRegistrationWorker.submit(() -> withThreadName("ResourceResolverFactory deregistration",
+                this::doUnregisterFactory));
+    }
+
+    /**
+     * Register the factory.
+     */
+    private void doRegisterFactory() {
+        LOG.debug("performing registerFactory, factoryRegistration == {}", factoryRegistration);
+        if (this.factoryRegistration == null) {
+            this.factoryRegistration = new FactoryRegistration(activator.getBundleContext(), activator);
+        }
+        LOG.debug("finished performing registerFactory, factoryRegistration == {}", factoryRegistration);
+    }
+
+    /**
+     * Unregister the factory (if registered).
+     */
+    private void doUnregisterFactory() {
+        LOG.debug("performing unregisterFactory, factoryRegistration == {}", factoryRegistration);
+        if (this.factoryRegistration != null) {
+            this.factoryRegistration.unregister();
+            this.factoryRegistration = null;
+        }
+        LOG.debug("finished performing unregisterFactory, factoryRegistration == {}", factoryRegistration);
+    }
+
+    private static void withThreadName(String threadName, Runnable task) {
+        final String name = Thread.currentThread().getName();
```

Comment:
Are you trying to expose what is going on via the thread name? It might be better for observability to keep a stable thread name prefix and reflect the current task in the second part, e.g. `getClass().getSimpleName() + " - " threadName`

Reply:
I found it helpful to have the registration/deregistration reflected in the thread-name, because service registrations trigger further service registrations etc. So it helps understand, which changes in the system happened due to the registration/deregistration.

Regarding the naming of the thread, yes the class-name as a prefix would work. De-facto, I am using the prefix "ResourceResolverFactory". I can incorporate that into the method, if you think it would be valuable.

Reply:
Ah, that works for me. Maybe extract the prefix as a constant value somewhere, just to make it clear?

Reply:
Sure 👍

Reply:
Fixed in https://github.com/apache/sling-org-apache-sling-resourceresolver/pull/100/commits/b28d4d0b208ea29579566f0b9bd823a1a3933af3